mentat/code_context.py

```diff
@@ -147,188 +60,313 @@ class CodeContextSettings:
-            file_message.append(posix_rel_path)
-
-            file_lines = all_file_lines[rel_path]
-            for i, line in enumerate(file_lines, start=1):
-                if file.contains_line(i):
-                    if provide_line_numbers:
-                        file_message.append(f"{i}:{line}")
-                    else:
-                        file_message.append(f"{line}")
-            file_message.append("")
-
-            if rel_path in self.diff_context.files:
-                file_message = self.diff_context.annotate_file_message(
-                    rel_path, file_message
-                )
-
-            code_message += file_message
-
-        if self.code_map is not None:
-            code_message_tokens = count_tokens("\n".join(code_message), model)
-            context_size = model_context_size(model)
-            if context_size:
-                max_tokens_for_code_map = context_size - code_message_tokens
-                if self.code_map.token_limit:
-                    code_map_message_token_limit = min(
-                        self.code_map.token_limit, max_tokens_for_code_map
-                    )
-                else:
-                    code_map_message_token_limit = max_tokens_for_code_map
-            else:
-                code_map_message_token_limit = self.code_map.token_limit
 
-            code_map_message = await self.code_map.get_message(
-                token_limit=code_map_message_token_limit
-            )
-            if code_map_message:
-                match (code_map_message.level):
-                    case "signatures":
-                        level_message = "full syntax tree"
-                    case "no_signatures":
-                        level_message = "partial syntax tree"
-                    case "filenames":
-                        level_message = "filepaths only"
-
-                message = f"\nIncluding CodeMap ({level_message})"
-                await stream.send(message, color="green")
-                code_message += [f"\n{code_map_message.content}"]
-            else:
-                message = [
-                    "\nExcluding CodeMap from system message.",
-                    "Reason: not enough tokens available in model context.",
-                ]
-                message = "\n".join(message)
-                await stream.send(message, color="yellow")
+        features = self._get_include_features()
+        include_feature_tokens_task = [f.count_tokens(model) for f in features]
+        results = await asyncio.gather(*include_feature_tokens_task)
+        include_feature_tokens = sum(results) + count_tokens(
+            "\n".join(code_message), model
+        )
+        _max_auto = max(0, max_tokens - include_feature_tokens)
+        _max_user = self.settings.auto_tokens
+        if _max_auto == 0 or _max_user == 0:
+            self.features = features
+        else:
+            auto_tokens = _max_auto if _max_user is None else min(_max_auto, _max_user)
```

Comment:
I hate this pattern and did something similar (but worse) [here](https://github.com/AbanteAI/mentat/blob/3c4db32fcc83c66fb2a886d8f0d94b3c53d27c3c/mentat/conversation.py#L79). I wonder if we should define our own `safe_min` and `safe_max` functions in util that return the min/max non None parameter and have sensible default returns like 0/infinity.

Reply:
Haha why do you hate it? Seems concise, and does something pretty specific.

Reply:
Yeah it's fine. I guess I just want min/max to accept `None` values and be sane about it. But it's probably better that they don't as that could mask other bugs.